File: rag-evaluation-backend/app/schemas/dataset.py
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from uuid import UUID

# 数据集类型枚举
# standard: 普通检索数据集 - 以直接事实检索为主(easy/medium，事实型/概念型)
# advanced: 高级检索数据集 - 含推理、归纳能力的问题(medium/hard，推理型/归纳型/比较型)
DATASET_TYPE_STANDARD = "standard"
DATASET_TYPE_ADVANCED = "advanced"

# 基础数据集模型
class DatasetBase(BaseModel):
    name: str
    description: Optional[str] = None
    is_public: Optional[bool] = False
    tags: Optional[List[str]] = Field(default_factory=list)
    dataset_metadata: Optional[Dict[str, Any]] = Field(default_factory=dict)
    dataset_type: Optional[str] = DATASET_TYPE_STANDARD  # standard | advanced

# 创建数据集时的请求模型
class DatasetCreate(DatasetBase):
    pass

# 更新数据集时的请求模型
class DatasetUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_public: Optional[bool] = None
    tags: Optional[List[str]] = None
    dataset_metadata: Optional[Dict[str, Any]] = None

# 数据集响应模型
class DatasetOut(DatasetBase):
    id: str
    user_id: Optional[str] = None
    question_count: int = 0
    dataset_type: str = DATASET_TYPE_STANDARD
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

# 数据集详情响应模型
class DatasetDetail(DatasetOut):
    projects: List[Dict[str, str]] = Field(default_factory=list)
    
    model_config = ConfigDict(from_attributes=True)

# 项目-数据集关联模型
class ProjectDatasetLink(BaseModel):
    project_id: str
    dataset_id: str

    model_config = ConfigDict(from_attributes=True)

# 批量关联数据集请求
class BatchLinkDatasets(BaseModel):
    dataset_ids: List[str] 

File: rag-evaluation-frontend/src/types/dataset.ts
/**
 * 数据集类型定义
 *
 * standard - 普通检索数据集
 *   以直接事实检索为主，问题答案可从单一文本段落中直接找到。
 *   难度分布以 easy/medium 为主，问题类型以事实型(factoid)、概念型(conceptual)为主。
 *   用途：测试 RAG 系统基础检索准确性。
 *
 * advanced - 高级检索数据集
 *   包含需要推理、归纳或跨文档整合的问题，答案无法直接从单一片段获取。
 *   难度分布以 medium/hard 为主，问题类型包含推理型(reasoning)、归纳型(inferential)、比较型(comparative)。
 *   用途：测试 RAG 系统的深度理解、多跳检索和推理能力。
 */
export type DatasetType = 'standard' | 'advanced';

export interface Dataset {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  is_public: boolean;
  tags: string[];
  dataset_metadata: Record<string, any>;
  /** 数据集类型: standard=普通检索, advanced=高级检索(含推理/归纳) */
  dataset_type: DatasetType;
  question_count: number;
  created_at: string;
  updated_at: string;
}

export interface DatasetDetail extends Dataset {
  projects: Array<{
    id: string;
    name: string;
  }>;
}

export interface Question {
  id: string;
  dataset_id: string;
  question_text: string;
  standard_answer: string;
  category?: string;
  difficulty?: string;
  type?: string;
  tags?: Record<string, any>;
  question_metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface CreateDatasetRequest {
  name: string;
  description?: string;
  is_public: boolean;
  tags?: string[];
  dataset_metadata?: Record<string, any>;
  dataset_type?: DatasetType;
}

export interface UpdateDatasetRequest {
  name?: string;
  description?: string;
  is_public?: boolean;
  tags?: string[];
  dataset_metadata?: Record<string, any>;
  dataset_type?: DatasetType;
}

export interface DatasetListParams {
  page?: number;
  size?: number;
  search?: string;
  tag?: string;
  is_public?: boolean;
}

export interface ImportDataRequest {
  dataset_id: string;
  file?: File;
  questions?: Array<{
    question_text: string;
    standard_answer: string;
    category?: string;
    difficulty?: string;
    tags?: Record<string, any>;
  }>;
}

export interface ProjectDatasetLink {
  project_id: string;
  dataset_id: string;
}

export interface BatchLinkDatasetsRequest {
  project_id: string;
  dataset_ids: string[];
}

export interface RagAnswer {
  id: string;
  question_id: string;
  answer_text: string;
  source_type: string;
  version: string;
  api_config_id?: string;
  response_time?: number;
  token_count?: number;
  dataset_metadata?: any;
  created_at: string;
  updated_at: string;
}

export interface CreateRagAnswerRequest {
  question_id: string;
  answer_text: string;
  source_type: string;
  version?: string;
  api_config_id?: string;
  response_time?: number;
  token_count?: number;
  dataset_metadata?: any;
}

export interface ImportQuestionWithRagRequest {
  dataset_id: string;
  questions: {
    question_text: string;
    standard_answer: string;
    category?: string;
    difficulty?: string;
    type?: string;
    tags?: string[];
    rag_answer?: CreateRagAnswerRequest;
  }[];
} 